keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,426 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP aims to introduce node-affinity-based eviction.
+In this KEP, we will introduce a new nodeAffinity type: `requiredDuringSchedulingRequiredDuringExecution`.
+A new controller called `node-affinity-eviction` will be added to the controller-manager. 
+This controller will monitor changes in node labels and evict pods when their `requiredDuringSchedulingRequiredDuringExecution` node affinity is no longer met.
+
+## Motivation
+
+- Enable `requiredDuringSchedulingRequiredDuringExecution` node affinity for pod scheduling and execution.
+- Provide users with greater control over pod placement.
+- Ensure predictable and consistent pod behavior in various scenarios.
+- Address use cases with specific placement and execution requirements.
+  
+### Goals
+
+- Introduce the `RequiredDuringSchedulingRequiredDuringExecution` **nodeAffinity** type.
+- Add `node-affinity-eviction` controller to ensure pods being evicted if the selectors are no longer met at runtime.
+
+### Non-Goals
+
+- This KEP does not aim to introduce the `RequiredDuringSchedulingRequiredDuringExecution` **podAffinity** type and **podAntiAffinity** type.
+- This feature does not respect PodDisruptionBudget during the eviction.
+
+## Proposal
+- Implement the `requiredDuringSchedulingRequiredDuringExecution` node affinity type.
+- Update the kube-scheduler, kube-controller-manager and kube-apiserver.
+- Define the behavior, syntax, and use cases in documentation.
+- Facilitate testing, adoption, and maintenance of this feature.
+
+### User Stories (Optional)
+
+#### Optimizing Workloads Migrating
+As an engineer of an AI-platform, we have different deployments for inference/training. 
+Our different deployments require different types of gpu.
+For example, some can only run on nodes with `gpu-type-name=T4`, and some can only run on nodes with `gpu-type-name=P100`.
+There are also some deployments which can run on both nodes(They may have made adaptations or they may not need a GPU at all).
+
+Sometimes, our customers want to add some new powerful machines to the cluster.
+For example, some machines with gpu type `T4`, but better ssd and more cpu are added to the cluster.
+As a result, they want to migrate old deployments that must use T4 GPUs to these powerful new machines. 
+As for the old machines, they can continue to run some services that do not require `T4` GPUs.
+
+With this feature, I can simply remove `gpu-type-name=T4` label from the old node and adds it to the new nodes, so I can move target deployments and keep the remaining pods intact.
```

Comment:
I still believe this is the wrong use case. Your story says: "they want to migrate old deployments", but then you change the nodes, meaning that this is an action done by a cluster administrator.

It would be more appropriate for a user to add a nodeSelector for something like "ssd=true". Then they control the timing of the migration.

Or, if it has to be an administrator operation, the administrator would add a taint to the original nodepool, something like "legacy=true".

I want to support this feature, but your use case is unfortunately not sound.

Reply:
labels generally represent a characteristic of the node. The original nodes still have a GPU T4, so it's not appropriate to change it.

Reply:
> It would be more appropriate for a user to add a nodeSelector for something like "ssd=true".

In the long run, yes. But if none of the current deployments have node-affinity "ssd=true". It will be more convenient to change the node's label to achieve this purpose(migration).
Yea, I agree it sounds like there's a better solution at the first place...

> I want to support this feature, but your use case is unfortunately not sound.

So, story 1 is what we actually encountered in real life, now it looks like some historical debt.
What about story 2? Is it a valid case to you ? @alculquicondor

Reply:
> labels generally represent a characteristic of the node. The original nodes still have a GPU T4, so it's not appropriate to change it.

So, `gpu-type-name=T4` sounds like it tells the gpu type of the node. But in our case, I think it's more like `allow-services-requiresT4-runs=true`, meaning this node allow services which requires gpu type T4 runs on it.
And in our case, it's appropriate to change the old machine, as it is considered obsolete by us, we don't want our main services running on it, but we still want it to continue running some less important services.

Reply:
> And in our case, it's appropriate to change the old machine, as it is considered obsolete by us.

Again, why not use a taint to mark obsolete nodes instead?

Reply:
That would be we didn't add corresponding tolerations to deployments before🤦.
I agree this is more like a fix for the inappropriate use of scheduling by users, rather than a common scenario.

I'll make `Multi Tenant` the only user case.

Reply:
You don't have to have tolerations in your deployments. That's actually the point. Because you don't have tolerations, your Pods would be evicted and they would be ok to run in the new nodes.

Reply:
There are also deployments which we don't want to be moved from old machines, so we have to add toleration for those deployments.
But let's get over with this as this story is already removed from kep😂.